### ISSUE TITLE
Zone.js standardization track push

### ISSUE THREAD

**Issue by bryanrideshark** [angular/angular#48454]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Tl;dr;

I would like to see a push for Zone.js to be brought into the standards track with TC39, or at least Async context tracking (similar to node.js async hooks). I'm using Zone.js in production code to track async context variables, and we'll be in a bind if we lose this ability.

Can we push to get Zone.js standardized again, or at least something that would allow us to save state in the async execution context?

# Backstory

In our organization, we have a concept called "Flows". It's similar to ngrx actions, but they are more like a function with a declared return type. I tell new employees they can think about it like a C header file, that gets linked at runtime; It uses the Angular dependency injector under the hood to abstract away a lot of configuration snafus.

We use flows for a lot of things, including navigation between sections of our app. A real world example is the "Go To Dashboard" function - each user depending on their individual attributes might have a different dashboard. This is complicated further by our large list of clients. Each of our clients can reconfigure their instance to change what "Dashboard" means for their organization. Sometimes, these dashboards are a 3rd URL that is _not_ controlled by Angular, or at least not the same Angular app.

Using a "flow", we can easily abstract away the implementation details of figuring out _which_ dashboard should be rendered, and how to get there. All navigation links can then just invoke the flow, eventually navigating the user.

While this can be invoked with a (click) callback, I like anchor tags, and would like to use these same flows to determine a static href, not just navigate. Simple enough; I can just flip the switch on the navigation engine we use to "record" rather than actually navigate there.

The problem comes down to `async`. Because the same flow could be triggered programmatically (in which case we want to actually navigate there) at the same time the href determination service is running, I need a way to know the execution context at the time the function is executing.

Zone.js works great for this! Since I can just fork the zone before executing, I can know whether or not the current callback should be _executing_ or _recording_. I use WeakMaps to track settings per zone, using Zone.current to look up the appropriate behavior inside the callback.

## My probelm

If Angular moves to a Zoneless future (because of native async / await), I will lose a major part of my capabilities, and have to maintain two separate execution paths for href links, vs button links.

Are there any other ways to resolve this, without Zone.js? How can we track async execution context without it?

### Proposed solution

I could live without Zone.js but it would require two separate codepaths any time you needed something you could previously sandbox with Zone.

### Alternatives considered

I don't think we have any alternatives unless async context tracking is standardized.

**Comment by alxhub** [comment]
Hi @bryanrideshark,

As you mentioned, we're exploring making zone.js optional for Angular (which involves a significant update to Angular's reactivity model). The motivations for this are numerous (you can watch @jelbourn and my [talk](https://www.youtube.com/watch?v=IY-QOz4oLCE) from ng-conf this year for an overview). zone.js issues with async/await are only a minor part of this equation.

In the past we've made efforts towards standardization from zone.js, and encountered significant resistance. Given that, and the fact that zone optionality is a part of our future strategy, we're not going to invest in standardization of zones.

That said, there is no reason why you can't continue using zones in your application, even if the application is migrated to not use zones for change detection.

**Comment by bryanrideshark** [comment]
It seems like there is a push in the standards community to make Node's `async_hooks` library standardized. It would solve my use case.

When can we get an RFC for the reactivity changes to Angular?

**Comment by alxhub** [comment]
Reactivity is one of our main focuses for 2023, so I would expect an RFC sometime next year.
